zetaclient/chains/evm/observer/v2_outbound.go

```diff
@@ -0,0 +1,256 @@
+package observer
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strings"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/pkg/errors"
+	"github.com/zeta-chain/protocol-contracts/v2/pkg/erc20custody.sol"
+	"github.com/zeta-chain/protocol-contracts/v2/pkg/gatewayevm.sol"
+
+	"github.com/zeta-chain/zetacore/pkg/chains"
+	crosschaintypes "github.com/zeta-chain/zetacore/x/crosschain/types"
+	"github.com/zeta-chain/zetacore/zetaclient/chains/evm"
+)
+
+// ParseOutboundEventV2 parses an event from an outbound with protocol contract v2
+func ParseOutboundEventV2(
+	cctx *crosschaintypes.CrossChainTx,
+	receipt *ethtypes.Receipt,
+	transaction *ethtypes.Transaction,
+	custodyAddr ethcommon.Address,
+	custody *erc20custody.ERC20Custody,
+	gatewayAddr ethcommon.Address,
+	gateway *gatewayevm.GatewayEVM,
+) (*big.Int, chains.ReceiveStatus, error) {
+	// return failed status if receipt status is failed
+	if receipt.Status == ethtypes.ReceiptStatusFailed {
+		return big.NewInt(0), chains.ReceiveStatus_failed, nil
+	}
+
+	outboundType := evm.ParseOutboundTypeFromCCTX(*cctx)
+	switch outboundType {
+	case evm.OutboundTypeGasWithdraw:
+	case evm.OutboundTypeGasWithdrawRevert:
+		// simple transfer, no need to parse event
+		return transaction.Value(), chains.ReceiveStatus_success, nil
+	case evm.OutboundTypeERC20Withdraw:
+	case evm.OutboundTypeERC20WithdrawRevert:
+		return ParseAndCheckERC20CustodyWithdraw(cctx, receipt, custodyAddr, custody)
+	case evm.OutboundTypeERC20WithdrawAndCall:
+		return ParseAndCheckERC20CustodyWithdrawAndCall(cctx, receipt, custodyAddr, custody)
+	case evm.OutboundTypeGasWithdrawAndCall:
+	case evm.OutboundTypeCall:
+		// both gas withdraw and call and no-asset call uses gateway execute
+		// no-asset call simply hash msg.value == 0
+		return ParseAndCheckGatewayExecuted(cctx, receipt, gatewayAddr, gateway)
+	case evm.OutboundTypeGasWithdrawRevertAndCallOnRevert:
+	case evm.OutboundTypeERC20WithdrawRevertAndCallOnRevert:
+		return ParseAndCheckGatewayReverted(cctx, receipt, gatewayAddr, gateway)
+	}
+	return big.NewInt(0), chains.ReceiveStatus_failed, fmt.Errorf("unsupported outbound type %d", outboundType)
+}
+
+// ParseAndCheckGatewayExecuted parses and checks the gateway execute event
+func ParseAndCheckGatewayExecuted(
+	cctx *crosschaintypes.CrossChainTx,
+	receipt *ethtypes.Receipt,
+	gatewayAddr ethcommon.Address,
+	gateway *gatewayevm.GatewayEVM,
+) (*big.Int, chains.ReceiveStatus, error) {
+	params := cctx.GetCurrentOutboundParam()
+
+	for _, vLog := range receipt.Logs {
+		executed, err := gateway.GatewayEVMFilterer.ParseExecuted(*vLog)
+		if err == nil {
```

Comment:
We can perform here `if err != nil { continue }` and then execute the rest of the logic, to avoid identation.